File: output/playwright/nested-wires/capture-m4.mjs
/** Headless evidence on the existing Vite; no layout or DOM geometry mutation. */
const directory = 'output/playwright/nested-wires/';
export async function capture(page) {
  await page.setViewportSize({ width: 1920, height: 1440 });
  const loads = [];
  for (let i = 0; i < 5; i += 1) {
    await page.goto('http://127.0.0.1:5188/roads-prototype.html?nested');
    await page.waitForFunction(
      () => performance.getEntriesByName('roads:navigation-to-ready').length > 0,
    );
    loads.push(
      await page.evaluate(() =>
        Object.fromEntries(
          performance
            .getEntriesByType('measure')
            .filter((e) => e.name.startsWith('roads:'))
            .map((e) => [e.name, e.duration]),
        ),
      ),
    );
  }
  await page.waitForTimeout(250);
  await page.screenshot({ path: `${directory}m4-overview.png` });
  const checkbox = page.getByRole('checkbox', { name: 'Show roads' });
  await checkbox.uncheck();
  await page.screenshot({ path: `${directory}m4-roads-off.png` });
  const roadsOff = !(await checkbox.isChecked());
  await checkbox.check();
  const hub = page.locator('[data-node-id="node-23"]');
  const bounds = await hub.boundingBox();
  await page.mouse.move(bounds.x + bounds.width / 2, bounds.y + bounds.height / 2);
  await page.mouse.wheel(0, -600);
  await page.waitForTimeout(350);
  await page.screenshot({ path: `${directory}m4-hub-closeup.png` });
  return {
    loads,
    medianMilliseconds: loads
      .map((l) => l['roads:navigation-to-ready'])
      .toSorted((a, b) => a - b)[2],
    screenshots: ['m4-overview', 'm4-roads-off', 'm4-hub-closeup'],
    roadsOff,
    wires: await page.locator('[data-wire-id]').count(),
    nodes: await page.locator('[data-node-id]').count(),
    layoutRecalcCount: await page.evaluate(() => window.__layoutRecalcCount),
  };
}

File: output/playwright/nested-wires/capture-m4.py
"""Reproduce Chromium evidence through the installed Playwright CLI skill. Never starts Vite."""
import json
import subprocess
from pathlib import Path

ROOT = Path(__file__).resolve().parents[3]
EVIDENCE = ROOT / 'output/playwright/nested-wires'
CLI = Path.home() / '.codex/skills/playwright/scripts/playwright_cli.sh'


def command(*args):
    result = subprocess.run([str(CLI), '-s=m4-capture', *args], cwd=ROOT,
                            text=True, capture_output=True)
    if result.returncode: raise RuntimeError(result.stdout + result.stderr)
    if '### Error' in result.stdout:
        raise RuntimeError(result.stdout)
    return result.stdout


command('open', 'http://127.0.0.1:5188/roads-prototype.html?nested')
command('snapshot')
source = (EVIDENCE / 'capture-m4.mjs').read_text().replace('export async function', 'async function')
output = command('run-code', 'async () => {\n' + source + '\nreturn capture(page);\n}')
report = json.loads(output.split('### Result\n', 1)[1].split('\n###', 1)[0])
(EVIDENCE / 'm4-browser.json').write_text(json.dumps(report, indent=2) + '\n')
print(f"PASS M4 five loads={[l['roads:navigation-to-ready'] for l in report['loads']]}; median={report['medianMilliseconds']:.3f} ms; M2=243.7 ms; ceiling=300 ms")
assert report['medianMilliseconds'] <= 300, report
assert report['wires'] == 26 and report['nodes'] == 24 and report['layoutRecalcCount'] == 1
print('PASS 24 nodes, 26 wires, one layout calculation; three M4 screenshots captured')

assert report['roadsOff']
command('close')
